src/_gettsim/interface.py

```diff
@@ -2,661 +2,453 @@
-        column.
-    targets : list of str
-        List of strings with names of functions whose output is actually
-        needed by the user. By default, ``targets`` contains all key outputs as
-        defined by `gettsim.config.DEFAULT_TARGETS`.
-    columns_overriding_functions : list of str
-        Names of columns in the data which are preferred over function defined in the
-        tax and transfer system.
-    check_minimal_specification : {"ignore", "warn", "raise"}, default "ignore"
-        Indicator for whether checks which ensure the most minimal configuration should
-        be silenced, emitted as warnings or errors.
-
-    Returns
-    -------
-    dag : networkx.DiGraph
-        The DAG of the tax and transfer system.
-
-    """
-    # Create DAG and perform checks which depend on data which is not part of the DAG
-    # interface.
-
-    dag = dags.dag.create_dag(
-        functions=all_functions,
-        targets=targets,
+    input_structure = dags.create_input_structure_tree(
+        functions_tree_not_overridden,
     )
-    _fail_if_columns_overriding_functions_are_not_in_dag(
-        dag, columns_overriding_functions, check_minimal_specification
-    )
-
-    return dag
 
+    # Remove unnecessary elements from user-provided data.
+    input_data_tree = _create_input_data_for_concatenated_function(
+        data_tree=data_tree_with_correct_types,
+        functions_tree=functions_tree_with_partialled_parameters,
+        targets_tree=targets_tree,
+        input_structure=input_structure,
+    )
 
-def _process_and_check_data(data):
-    """Process data and perform several checks.
+    _fail_if_group_variables_not_constant_within_groups(input_data_tree)
+    _fail_if_foreign_keys_are_invalid(
+        data_tree=input_data_tree,
+        p_ids=data_tree_with_correct_types.get("groupings", {}).get("p_id", {}),
+    )
 
-    Parameters
-    ----------
-    data : pandas.Series or pandas.DataFrame or dict of pandas.Series
-        Data provided by the user.
+    tax_transfer_function = dags.concatenate_functions_tree(
+        functions=functions_tree_with_partialled_parameters,
+        targets=targets_tree,
+        input_structure=input_structure,
+        name_clashes="raise",
+    )
 
-    Returns
-    -------
-    data : dict of pandas.Series
+    results = tax_transfer_function(input_data_tree)
 
-    """
-    if isinstance(data, pd.DataFrame):
-        _fail_if_duplicates_in_columns(data)
-        data = dict(data)
-    elif isinstance(data, pd.Series):
-        data = {data.name: data}
-    elif isinstance(data, dict) and all(
-        isinstance(i, pd.Series) for i in data.values()
-    ):
-        pass
-    else:
-        raise NotImplementedError(
-            "'data' is not a pd.DataFrame or a pd.Series or a dictionary of pd.Series."
+    if debug:
+        results = upsert_tree(
+            base_tree=results,
+            update_tree=data_tree_with_correct_types,
```

Comment:
```suggestion
            to_upsert=data_tree_with_correct_types,
```
A bit pedantic, but let's match the function definition here.